### PR TITLE
Adding error messages to contracts

### LINKED ISSUE #934: Users SBAT see explanations of failed validation checks on requires
methods in:
-[] LockedGold
-[] Validators
-[] Governance
-[] Attestations
-[] Escrow

### PR DESCRIPTION
**Pull request by mrsmkl**
### Description

Added error messages, otherwise the behaviour shouldn't change.

### Tested

Deployment cost of Validators contract grew to 8258626 gas from 7315248 gas.

### Other changes

Also removed some compiler warnings.

### Related issues

- Fixes #934

### CODE REVIEW

File: packages/protocol/contracts/governance/Governance.sol [hunk 1]
```diff
@@ -351,8 +362,12 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
   function setParticipationBaseline(uint256 participationBaseline) public onlyOwner {
     FixidityLib.Fraction memory participationBaselineFrac = FixidityLib.wrap(participationBaseline);
     require(
-      FixidityLib.isProperFraction(participationBaselineFrac) &&
-      !participationBaselineFrac.equals(participationParameters.baseline)
+      FixidityLib.isProperFraction(participationBaselineFrac),
+      "Improper fraction"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: `Participation baseline greater than one`

File: packages/protocol/contracts/governance/Governance.sol [hunk 2]
```diff
@@ -365,8 +380,12 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
   function setParticipationFloor(uint256 participationFloor) public onlyOwner {
     FixidityLib.Fraction memory participationFloorFrac = FixidityLib.wrap(participationFloor);
     require(
-      FixidityLib.isProperFraction(participationFloorFrac) &&
-      !participationFloorFrac.equals(participationParameters.baselineFloor)
+      FixidityLib.isProperFraction(participationFloorFrac),
+      "Improper fraction"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: `Participation floor greater than one`

File: packages/protocol/contracts/governance/Governance.sol [hunk 3]
```diff
@@ -379,8 +398,12 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
   function setBaselineUpdateFactor(uint256 baselineUpdateFactor) public onlyOwner {
     FixidityLib.Fraction memory baselineUpdateFactorFrac = FixidityLib.wrap(baselineUpdateFactor);
     require(
-      FixidityLib.isProperFraction(baselineUpdateFactorFrac) &&
-      !baselineUpdateFactorFrac.equals(participationParameters.baselineUpdateFactor)
+      FixidityLib.isProperFraction(baselineUpdateFactorFrac),
+      "Improper fraction"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: `Baseline update factor greater than one`

File: packages/protocol/contracts/governance/Governance.sol [hunk 4]
```diff
@@ -393,8 +416,12 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
   function setBaselineQuorumFactor(uint256 baselineQuorumFactor) public onlyOwner {
     FixidityLib.Fraction memory baselineQuorumFactorFrac = FixidityLib.wrap(baselineQuorumFactor);
     require(
-      FixidityLib.isProperFraction(baselineQuorumFactorFrac) &&
-      !baselineQuorumFactorFrac.equals(participationParameters.baselineQuorumFactor)
+      FixidityLib.isProperFraction(baselineQuorumFactorFrac),
+      "Improper fraction"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: `Baseline quorum factor greater than one`

File: packages/protocol/contracts/governance/Governance.sol [hunk 5]
```diff
@@ -634,12 +674,19 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
...
+    );
     Proposals.Stage stage = proposal.getDequeuedStage(stageDurations);
     bool expired = isDequeuedProposalExpired(proposal, stage);
     if (!expired) {
       // TODO(asa): Think through the effects of changing the passing function
-      require(stage == Proposals.Stage.Execution && _isProposalPassing(proposal));
+      require(
+        stage == Proposals.Stage.Execution &&
+        _isProposalPassing(proposal),
+        "Proposal not in execution stage"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Governance.sol]: `Proposal not in execution stage or not passing`

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 6]
```diff
@@ -229,15 +233,17 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     nonReentrant
   {
     // TODO: split and add error messages for better dev feedback
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: Remove todo

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 7]
```diff
@@ -262,10 +268,15 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     payable
     returns (uint256)
   {
-    require(isAccount(msg.sender) && !isVoting(msg.sender));
+    require(
+      isAccount(msg.sender) &&
+      !isVoting(msg.sender),
+      "Only non-voting accounts can add commitment"
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: nit: "commitments"

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 8]
```diff
@@ -319,13 +332,15 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     nonReentrant
     returns (uint256)
   {
-    require(isAccount(msg.sender) && !isVoting(msg.sender));
+    require(isAccount(msg.sender), "extendCommitment: Unknown account");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: Remove "extendCommitment: " for consistency

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 9]
```diff
@@ -319,13 +332,15 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
...
+    require(!isVoting(msg.sender), "extendCommitment: Account locked");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: Ditto

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 10]
```diff
@@ -319,13 +332,15 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
...
     // solhint-disable-next-line not-rely-on-time
-    require(availabilityTime > now);
+    require(availabilityTime > now, "Cannot extend commitment to past time");
     // _redeemRewards(msg.sender);
     Account storage account = accounts[msg.sender];
     Commitment storage notified = account.commitments.notified[availabilityTime];
-    require(notified.value >= value && value > 0);
+    require(notified.value >= value, "Attempted value larger than committed");
+    require(value > 0, "extendCommitment: Value zero");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: Ditto

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 11]
```diff
@@ -319,13 +332,15 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     nonReentrant
     returns (uint256)
   {
-    require(isAccount(msg.sender) && !isVoting(msg.sender));
+    require(isAccount(msg.sender), "extendCommitment: Unknown account");
+    require(!isVoting(msg.sender), "extendCommitment: Account locked");
     // solhint-disable-next-line not-rely-on-time
-    require(availabilityTime > now);
+    require(availabilityTime > now, "Cannot extend commitment to past time");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: "Cannot extend available notified commitment"

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 12]
```diff
@@ -347,19 +362,20 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     nonReentrant
     returns (uint256)
   {
-    require(isAccount(msg.sender) && !isVoting(msg.sender));
+    require(isAccount(msg.sender), "withdrawCommitment: Unknown account");
+    require(!isVoting(msg.sender), "withdrawCommitment: Account locked");
     // _redeemRewards(msg.sender);
     // solhint-disable-next-line not-rely-on-time
-    require(now >= availabilityTime);
+    require(now >= availabilityTime, "Trying to withdraw too early");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: "Notified commitment not available"

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 13]
```diff
@@ -380,12 +396,14 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     nonReentrant
     returns (uint256)
   {
-    require(isAccount(msg.sender) && !isVoting(msg.sender));
+    require(isAccount(msg.sender), "increaseNoticePeriod: Unknown account");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: remove `increaseNoticePeriod: ` for consistency

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 14]
```diff
@@ -380,12 +396,14 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
...
+    require(!isVoting(msg.sender), "increaseNoticePeriod: Account locked");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: Ditto

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 15]
```diff
@@ -380,12 +396,14 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
...
     // _redeemRewards(msg.sender);
-    require(value > 0 && increase > 0);
+    require(value > 0, "Value zero");
+    require(increase > 0, "Increase zero");
     Account storage account = accounts[msg.sender];
     Commitment storage locked = account.commitments.locked[noticePeriod];
-    require(locked.value >= value);
+    require(locked.value >= value, "Attempted value larger than locked amount");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: `Value larger than commitment`

File: packages/protocol/contracts/governance/LockedGold.sol [hunk 16]
```diff
@@ -588,7 +609,7 @@ contract LockedGold is ILockedGold, ReentrancyGuard, Initializable, UsingRegistr
     if (value > 0) {
       address recipient = getDelegateFromAccountAndRole(_account, DelegateRole.Rewards);
       IERC20Token goldToken = IERC20Token(registry.getAddressFor(GOLD_TOKEN_REGISTRY_ID));
-      require(goldToken.transfer(recipient, value));
+      require(goldToken.transfer(recipient, value), "GLD transfer failed");
```
Comment (asaj) [inline packages/protocol/contracts/governance/LockedGold.sol]: `Transfer failed` for consistency

File: packages/protocol/contracts/governance/Proposals.sol [hunk 17]
```diff
@@ -247,7 +252,7 @@ library Proposals {
     view
     returns (uint256, address, bytes memory)
   {
-    require(index < proposal.transactions.length);
+    require(index < proposal.transactions.length, "getTransaction: bad index");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Proposals.sol:286]: nit: "Bad index" for consistency

File: packages/protocol/contracts/governance/Validators.sol [hunk 18]
```diff
@@ -179,7 +179,11 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
     external
     initializer
   {
-    require(_minElectableValidators > 0 && _maxElectableValidators >= _minElectableValidators);
+    require(
+      _minElectableValidators > 0 &&
+      _maxElectableValidators >= _minElectableValidators,
+      "Illegal number of validators"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: nit: s/Illegal/Invalid here and below

File: packages/protocol/contracts/governance/Validators.sol [hunk 19]
```diff
@@ -338,14 +345,25 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
       bytes(name).length > 0 &&
       bytes(url).length > 0 &&
       // secp256k1 public key + BLS public key + BLS proof of possession
-      publicKeysData.length == (64 + 48 + 96)
+      publicKeysData.length == (64 + 48 + 96),
+      "registerValidator: bad input"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: "Bad input"

File: packages/protocol/contracts/governance/Validators.sol [hunk 20]
```diff
@@ -373,7 +391,7 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
    */
   function deregisterValidator(uint256 index) external nonReentrant returns (bool) {
     address account = getAccountFromValidator(msg.sender);
-    require(isValidator(account));
+    require(isValidator(account), "deregisterValidator: Not a validator");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: "Not a validator"

File: packages/protocol/contracts/governance/Validators.sol [hunk 21]
```diff
@@ -392,7 +410,11 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
    */
   function affiliate(address group) external nonReentrant returns (bool) {
     address account = getAccountFromValidator(msg.sender);
-    require(isValidator(account) && isValidatorGroup(group));
+    require(
+      isValidator(account) &&
+      isValidatorGroup(group),
+      "Already affiliated"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Separate into two:
"Not a validator"
"Not a validator group"

File: packages/protocol/contracts/governance/Validators.sol [hunk 22]
```diff
@@ -409,9 +431,15 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
    */
   function deaffiliate() external nonReentrant returns (bool) {
     address account = getAccountFromValidator(msg.sender);
-    require(isValidator(account));
+    require(
+      isValidator(account),
+      "deaffiliate: not a validator"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: "Not a validator"

File: packages/protocol/contracts/governance/Validators.sol [hunk 23]
```diff
@@ -409,9 +431,15 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
...
+    );
     Validator storage validator = validators[account];
-    require(validator.affiliation != address(0));
+    require(
+      validator.affiliation != address(0),
+      "deaffiliate: not affiliated"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: "Not affiliated"

File: packages/protocol/contracts/governance/Validators.sol [hunk 24]
```diff
@@ -437,10 +465,19 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
     nonReentrant
     returns (bool)
   {
-    require(bytes(identifier).length > 0 && bytes(name).length > 0 && bytes(url).length > 0);
+    require(
+      bytes(identifier).length > 0 && bytes(name).length > 0 && bytes(url).length > 0,
+      "registerValidatorGroup: bad input"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: "Bad input"

File: packages/protocol/contracts/governance/Validators.sol [hunk 25]
```diff
@@ -437,10 +465,19 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
...
+    );
     address account = getAccountFromValidator(msg.sender);
-    require(!isValidator(account) && !isValidatorGroup(account));
-    require(meetsRegistrationRequirements(account, noticePeriods));
+    require(
+      !isValidator(account) && !isValidatorGroup(account),
+      "registerValidatorGroup: already registered"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Remove prefix here and everywhere

File: packages/protocol/contracts/governance/Validators.sol [hunk 26]
```diff
@@ -491,7 +535,11 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
    */
   function removeMember(address validator) external nonReentrant returns (bool) {
     address account = getAccountFromValidator(msg.sender);
-    require(isValidatorGroup(account) && isValidator(validator));
+    require(
+      isValidatorGroup(account) &&
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Separate into two:

"Not a group"
"Not a validator"

File: packages/protocol/contracts/governance/Validators.sol [hunk 27]
```diff
@@ -515,9 +563,13 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
     returns (bool)
   {
     address account = getAccountFromValidator(msg.sender);
-    require(isValidatorGroup(account) && isValidator(validator));
+    require(
+      isValidatorGroup(account) &&
+      isValidator(validator),
+      "Invalid validator or group"
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Ditto

File: packages/protocol/contracts/governance/Validators.sol [hunk 28]
```diff
@@ -515,9 +563,13 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
...
+    );
     ValidatorGroup storage group = groups[account];
-    require(group.members.contains(validator));
+    require(group.members.contains(validator), "Not contained in the group");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: "Validator not a member of the group"

File: packages/protocol/contracts/governance/Validators.sol [hunk 29]
```diff
@@ -899,7 +961,8 @@ contract Validators is IValidators, Ownable, ReentrancyGuard, Initializable, Usi
    */
   function _removeMember(address group, address validator) private returns (bool) {
     ValidatorGroup storage _group = groups[group];
-    require(validators[validator].affiliation == group && _group.members.contains(validator));
+    require(validators[validator].affiliation == group, "Not affiliated to group");
+    require(_group.members.contains(validator), "Not contained in the group");
```
Comment (asaj) [inline packages/protocol/contracts/governance/Validators.sol]: Not a member of the group

**Review by asaj — COMMENTED** [review]
Woohoo! Thanks for doing this, will make debugging life much easier

**Review by mcortesi — APPROVED** [review]
Thanks for doing this!!!

**Pull request merged**